MDEV-24197: Add option "innodb_force_recovery" for "mariabackup"

clickyotomy opened the pull request:
During the prepare phase of restoring backups, `mariabackup` does not seem to allow (or recognize) the option `innodb_force_recovery` for the embedded InnoDB server instance that it starts.

If page corruption observed during page recovery, the prepare step fails. While this is indeed the correct behavior ideally, allowing this option to be set in case of emergencies might be useful when the current backup is the only copy available. Some error messages during `--prepare` suggest to set `innodb_force_recovery` to `1`:

```
  [ERROR] InnoDB: Set innodb_force_recovery=1 to ignore corruption.
```

For backwards compatibility, `--innobackupex --apply-log` should also have this option.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1772) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1772) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1772) it.</sub>

dr-m commented:
I think that this is intended design. A failure to prepare a backup is a sign that something is seriously wrong. One should always prepare the backup immediately after making it, to ensure that everything went right. Hence, there should be no valid reason to use `innodb_force_recovery`. For `mariabackup --prepare`, only `innodb_force_recovery=1` might be applicable (and should not happen if everything works as intended, because `mariabackup --backup` is supposed to retry reading pages that carry an invalid checksum). Only for `mariabackup --prepare --export` one could argue that such a setting might be useful.

The bug that was stated as motivation for this looks like what was fixed in 549a70d7f01ce723b1b06fd7be569211a8270bcd.

clickyotomy commented:
Hello,

On 03/18/2021 12:39, Marko Mäkelä wrote:
> I think that this is intended design. A failure to prepare a backup is a sign that something is seriously wrong. One should always prepare the backup immediately after making it, to ensure that everything went right. Hence, there should be no valid reason to use `innodb_force_recovery`. For `mariabackup --prepare`, only `innodb_force_recovery=1` might be applicable (and should not happen if everything works as intended, because `mariabackup --backup` is supposed to retry reading pages that carry an invalid checksum). Only for `mariabackup --prepare --export` one could argue that such a setting might be useful.

You're right about `--backup`, and this option is not necessary
there. This was change intended for the "--prepare" stage only.
I forgot to check for that condition here.

Would adding a condition to check for `--prepare` and also limit
the `max_value` for that `my_option` to 1 if only `--prepare` is
passed (in case of emergencies) be acceptable?

For `--prepare --export`, I suppose the above condition would
apply here as well. Should the argument list for the `mariabackup`
command that runs inside `prepare_export()` also be updated to
`innodb_force_recovery`?
 
> The bug that was stated as motivation for this looks like what was fixed in 549a70d7f01ce723b1b06fd7be569211a8270bcd.

Yes, that was why I submitted this PR; thanks for addressing the issue!
Are the changes I mentioned above still necessary in that case?

Thanks.

-- 
Srinidhi Kaushik

dr-m commented:
Yes, I think that allowing `innodb_force_recovery=1` for `mariabackup --prepare` would make sense, for emergency scenarios, to be able to rescue some data in a case where log cannot be applied to a data page, typically due to some bug (that `mariabackup --backup` would be unable to detect).

I do not think that any larger values of the parameter would make any sense in `mariabackup`.

vaintroub commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -1570,6 +1571,11 @@ struct my_option xb_server_options[] =
+  {"innodb_force_recovery", OPT_INNODB_FORCE_RECOVERY,
+   "Crash recovery mode (for emergencies).",
+   (G_PTR*)&srv_force_recovery, (G_PTR*)&srv_force_recovery,
+   0, GET_UINT, OPT_ARG, 0, 0, 6, 0, 0, 0},
```
GET_ULONG. not GET_UINT

clickyotomy replied:
Updated; thanks!

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -1745,11 +1754,13 @@ static int prepare_export()
       " --defaults-file=./backup-my.cnf --defaults-group-suffix=%s --datadir=."
       " --innodb --innodb-fast-shutdown=0 --loose-partition"
       " --innodb_purge_rseg_truncate_frequency=1 --innodb-buffer-pool-size=%llu"
-      " --console  --log-error= --skip-log-bin --bootstrap  < "
+      " --console --log-error= --skip-log-bin --bootstrap"
+      " --innodb-force-recovery=%lu < "
       BOOTSTRAP_FILENAME IF_WIN("\"",""),
       mariabackup_exe,
       (my_defaults_group_suffix?my_defaults_group_suffix:""),
-      xtrabackup_use_memory);
+      xtrabackup_use_memory,
+      srv_force_recovery);
```
I think that we should output this special option only when it is set. Maybe, just add `"%s`" to the format string and pass `srv_force_recovery ? " --innodb-force-recovery=1" : ""` as the parameter.

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -1732,11 +1739,13 @@ static int prepare_export()
       " --defaults-extra-file=./backup-my.cnf --defaults-group-suffix=%s --datadir=."
       " --innodb --innodb-fast-shutdown=0 --loose-partition"
       " --innodb_purge_rseg_truncate_frequency=1 --innodb-buffer-pool-size=%llu"
-      " --console  --skip-log-error --skip-log-bin --bootstrap  < "
+      " --console --skip-log-error --skip-log-bin --bootstrap"
+      " --innodb-force-recovery=%lu < "
       BOOTSTRAP_FILENAME IF_WIN("\"",""),
-      mariabackup_exe, 
+      mariabackup_exe,
       orig_argv1, (my_defaults_group_suffix?my_defaults_group_suffix:""),
-      xtrabackup_use_memory);
+      xtrabackup_use_memory,
+      srv_force_recovery);
```
Do not output `--innodb-force-recovery=0` to keep compatible with older versions.

Side note: I think that @vaintroub could have made the `--defaults-file` branch unnecessary by including the English language strings directly in the executables, a while ago.

Side note: The `else` branch is using `sprintf` instead of the more appropriate `snprintf`. You could fix that as well.

clickyotomy replied:
I am not familiar with why the branch exists (if you would like me to remove it, please let me know). I will change that to `snprintf` though.

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -1549,6 +1550,12 @@ struct my_option xb_server_options[] =
+   "(for --prepare): Crash recovery mode (ignores page corruption; for"
+   " emergencies only).",
+   (G_PTR*)&srv_force_recovery, (G_PTR*)&srv_force_recovery,
+   0, GET_ULONG, OPT_ARG, 0, 0, SRV_FORCE_IGNORE_CORRUPT, 0, 0, 0},
```
We could add `compile_time_assert(SRV_FORCE_IGNORE_CORRUPT == 1)` somewhere.

clickyotomy replied:
I think we can add that in `innodb_init_param()`, right before the assignment.

dr-m reviewed (changes requested):
Thank you. This looks rather acceptable to me. I think that it would make sense in the 10.2 branch already.

There is also a file `extra/mariabackup/innobackupex.cc` that implements an alternative command line syntax, but we think that it is deprecated. It might make sense to support the option there as well.

It would be good to add a test case for this in `mysql-test/suite/mariabackup`, to demonstrate what happens if an out-of-bounds value is passed, or if the option is used outside `--prepare` (or `--apply-log` in the alternative command line syntax). And also to demonstrate that we do not write any `--innodb-force-recovery=0` in the basic case. Such backward compatibility is important if we apply this to the GA version branches.

@an3l Can you file an MDEV for this? The MDEV number and title should be the first line of the commit message.

@clickyotomy Please rebase this to the 10.2 branch.

clickyotomy commented:
@dr-m,

Here's a quick summary of changes:
 - Rebased onto `10.2`.
 - Added the option for `--innobackupex` (backwards compatibility).
 - Added `compile_time_assert(SRV_FORCE_IGNORE_CORRUPT == 1)` for `srv_force_recovery`.
 - The `snprintf` call in `prepare_export()` to not pass `--innodb-force-recovery=0`, when it is unspecified in the command line.
 - Changed `sprintf` to `snprintf` in `prepare_export()` (command line arguments for `--export` without defaults file).
 - Added tests for the new argument.

Thanks.

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -271,6 +271,9 @@ static char *xtrabackup_debug_sync = NULL;
 
 my_bool xtrabackup_incremental_force_scan = FALSE;
 
+/* Ignore currupt pages (disabled by default). */
+ulong xtrabackup_innodb_force_recovery = 0;
```
I think that we can just use the existing global variable `srv_force_recovery`.

clickyotomy replied:
Yes, I initially thought that it would work, but when testing locally, it doesn't seem to work if the configuration is set in `backup-my.cnf`. I think the command line argument overrides the value set in the configuration file (only in `--innobackupex` mode), and it probably happens here:
https://github.com/MariaDB/server/blob/c0ca3c4ffa45042d608e55326036443e8c4f600d/extra/mariabackup/xtrabackup.cc#L6532-L6538

Also, added a comment about that [here](https://github.com/MariaDB/server/pull/1772/files#diff-d2ff45ced531afaa25c4b8fd289f51d0de59a96907894dc5b13807cbbba31d74R2272-R2277).

dr-m replied:
@clickyotomy Thank you for clarifying this. I was not aware of this. @vlad-lesin @vaintroub do you have a comment? You have worked more on this code base than me.

However, I wonder if we could do with just one additional global variable that would be shared between `xtrabackup.cc` and `innobackupex.cc`.

clickyotomy replied:
@dr-m,
> However, I wonder if we could do with just one additional global variable that would be shared between `xtrabackup.cc` and `innobackupex.cc`.

Yes, that will work!

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -1985,6 +1998,11 @@ xb_get_one_option(int optid,
     ADD_PRINT_PARAM_OPT(innobase_buffer_pool_filename);
     break;
 
+  case OPT_INNODB_FORCE_RECOVERY:
+
+    ADD_PRINT_PARAM_OPT(xtrabackup_innodb_force_recovery);
+    break;
```
For backward compatibility, we would only want to display the parameter when it is set. Refer to the macro definition:
```c++
#define ADD_PRINT_PARAM_OPT(value)              \
  { \
    print_param_str << opt->name << "=" << value << "\n"; \
    param_set.insert(opt->name); \
  }
```
I think that this should be
```c++
  case OPT_INNODB_FORCE_RECOVERY:
    if (srv_force_recovery)
      ADD_PRINT_PARAM_OPT(srv_force_recovery);
    break;
```

clickyotomy replied:
Oh, OK. That makes sense, will update in the next iteration.

dr-m reviewed (changes requested):
Thank you for rebasing this to 10.2 and extending `innobackupex`. I’d suggest that we do not add 2 more global variables to duplicate the pre-existing `srv_force_recovery`.

Also, please make the first line of the commit message start with the string "MDEV-24197:", and remove the quotes of the error messages, because they are specific to 10.5, not 10.2. You could say: "Some error messages during --prepare suggest to set innodb_force_recovery=1."

Yes, it looks like this would address [MDEV-24197](https://jira.mariadb.org/browse/MDEV-24197). I found that file searching for open tickets before I would have filed one for this pull request.

clickyotomy commented:
@dr-m,

> Thank you for rebasing this to 10.2 and extending `innobackupex`. I’d suggest that we do not add 2 more global variables to duplicate the pre-existing `srv_force_recovery`.

About, `srv_force_recovery`, please check my comment [here](https://github.com/MariaDB/server/pull/1772#discussion_r603319978). If there isn't a need to have the configuration value read from the `backup-my.cnf` file in `--innobackupex --apply-log` mode, binding to `srv_force_recovery` would be sufficient. Since it works in `--prepare`, I thought it would be a necessary addition (for parity).

> Also, please make the first line of the commit message start with the string "MDEV-24197:", and remove the quotes of the error messages, because they are specific to 10.5, not 10.2. You could say: "Some error messages during --prepare suggest to set innodb_force_recovery=1."
>
> Yes, it looks like this would address [MDEV-24197](https://jira.mariadb.org/browse/MDEV-24197). I found that file searching for open tickets before I would have filed one for this pull request.

OK, I will update the commit message.

Thanks.

dr-m commented on extra/mariabackup/xtrabackup.cc:
```diff
@@ -2250,6 +2268,26 @@ innodb_init_param(void)
+	 * and for "--innobackupex". Since the command line argument is
+	 * parsed after the defaults file, it takes precedence.
+	 */
+	srv_force_recovery |= (ulong) xtrabackup_innodb_force_recovery;
```
If using an additional global variable is necessary, then this must be a normal assignment, that is,
```c++
	srv_force_recovery = xtrabackup_innodb_force_recovery;
```
The cast looks redundant (both variables are of the same type), and the `|=` could lead to surprising results if we later allowed other values of this parameter than 0 or 1. Maybe you wanted to write the following:
```c++
	if (xtrabackup_innodb_force_recovery) {
		srv_force_recovery = xtrabackup_innodb_force_recovery;
	}
```

clickyotomy replied:
You're right! Didn't think of cases other than `1`. Thanks.

clickyotomy commented:
@dr-m,
I think f1afd7d should address the review comments. Please let me know if there's anything else that needs to be updated.

Thanks.

clickyotomy commented:
@dr-m,

> @clickyotomy Thank you, the code changes looked good. But, the new test is crashing in the debug build:
> 
> ```
> #3  0x00007fb4e7477662 in __GI___assert_fail (
>     assertion=0x556f23c455eb "!srv_force_recovery", 
>     file=0x556f23c44af8 "/mariadb/10.2o/storage/innobase/srv/srv0start.cc", 
>     line=2325, 
>     function=0x556f23c44d6d "dberr_t innobase_start_or_create_for_mysql()")
>     at assert.c:101
> #4  0x0000556f2383985c in innobase_start_or_create_for_mysql ()
>     at /mariadb/10.2o/storage/innobase/srv/srv0start.cc:2325
> #5  0x0000556f2318fe21 in innodb_init ()
>     at /mariadb/10.2o/extra/mariabackup/xtrabackup.cc:2308
> #6  xtrabackup_prepare_func (argv=0x556f24d7ab50)
>     at /mariadb/10.2o/extra/mariabackup/xtrabackup.cc:6013
> ```
> 
> With this additional fix, it works for me:
> 
> ```diff
> diff --git a/storage/innobase/srv/srv0start.cc b/storage/innobase/srv/srv0start.cc
> index a7be1cdbf21..9548730b359 100644
> --- a/storage/innobase/srv/srv0start.cc
> +++ b/storage/innobase/srv/srv0start.cc
> @@ -2322,7 +2322,7 @@ innobase_start_or_create_for_mysql()
>  			to the data files and truncate or delete the log.
>  			Unless --export is specified, no further change to
>  			InnoDB files is needed. */
> -			ut_ad(!srv_force_recovery);
> +			ut_ad(srv_force_recovery <= SRV_FORCE_IGNORE_CORRUPT);
>  			ut_ad(srv_n_log_files_found <= 1);
>  			ut_ad(recv_no_log_write);
>  			buf_flush_sync_all_buf_pools();
> ```
> 
> ```
> mariabackup.innodb_force_recovery 'innodb' [ pass ]  32160
> ```
> 
> Can you please amend your commit with that? You could also update the date (`git commit --amend --date=now`).



Thanks for catching that. Added the change that you suggested, and tested locally; it passed.

```
mariabackup.innodb_force_recovery 'innodb' [ pass ]  28852
```

clickyotomy commented:
Thanks, @dr-m!
